File: frontend/dancee_app/lib/models/event.dart
import 'package:equatable/equatable.dart';
import 'venue.dart';
import 'event_info.dart';
import 'event_part.dart';

/// Represents a dance event with all its details.
///
/// This is the main model class for dance events in the application.
/// It contains all information about an event including venue, timing,
/// dance styles, additional information, and event parts.
///
/// This is an immutable class that uses Equatable for value equality comparison.
class Event extends Equatable {
  /// Unique identifier for the event
  final String id;

  /// The title of the event
  final String title;

  /// A detailed description of the event
  final String description;

  /// The name of the event organizer
  final String organizer;

  /// The venue where the event takes place
  final Venue venue;

  /// The start time of the event
  final DateTime startTime;

  /// The end time of the event
  final DateTime endTime;

  /// The duration of the event
  final Duration duration;

  /// List of dance styles featured at this event
  final List<String> dances;

  /// Additional information about the event (prices, URLs, etc.)
  final List<EventInfo> info;

  /// The parts/schedule of the event (workshops, parties, etc.)
  final List<EventPart> parts;

  /// Whether this event is marked as favorite by the user
  final bool isFavorite;

  /// Whether this event is in the past
  final bool isPast;

  /// Optional badge text (e.g., "TODAY", "IN 2 DAYS", "FINISHED")
  final String? badge;

  /// Creates an Event with all required fields and optional fields with defaults.
  const Event({
    required this.id,
    required this.title,
    required this.description,
    required this.organizer,
    required this.venue,
    required this.startTime,
    required this.endTime,
    required this.duration,
    required this.dances,
    this.info = const [],
    this.parts = const [],
    this.isFavorite = false,
    this.isPast = false,
    this.badge,
  });

  /// Creates a copy of this Event with the given fields replaced with new values.
  Event copyWith({
    String? id,
    String? title,
    String? description,
    String? organizer,
    Venue? venue,
    DateTime? startTime,
    DateTime? endTime,
    Duration? duration,
    List<String>? dances,
    List<EventInfo>? info,
    List<EventPart>? parts,
    bool? isFavorite,
    bool? isPast,
    String? badge,
  }) {
    return Event(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      organizer: organizer ?? this.organizer,
      venue: venue ?? this.venue,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      duration: duration ?? this.duration,
      dances: dances ?? this.dances,
      info: info ?? this.info,
      parts: parts ?? this.parts,
      isFavorite: isFavorite ?? this.isFavorite,
      isPast: isPast ?? this.isPast,
      badge: badge ?? this.badge,
    );
  }

  /// Converts this Event to a JSON map.
  /// Dates are serialized as ISO 8601 strings.
  /// Duration is serialized as total seconds (integer).
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'organizer': organizer,
      'venue': venue.toJson(),
      'startTime': startTime.toIso8601String(),
      'endTime': endTime.toIso8601String(),
      'duration': duration.inSeconds,
      'dances': dances,
      'info': info.map((i) => i.toJson()).toList(),
      'parts': parts.map((p) => p.toJson()).toList(),
      'isFavorite': isFavorite,
      'isPast': isPast,
      'badge': badge,
    };
  }

  /// Creates an Event from a JSON map.
  /// ISO 8601 date strings are converted back to DateTime objects.
  /// Duration in seconds is converted back to a Duration object.
  factory Event.fromJson(Map<String, dynamic> json) {
    return Event(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String,
      organizer: json['organizer'] as String,
      venue: Venue.fromJson(json['venue'] as Map<String, dynamic>),
      startTime: DateTime.parse(json['startTime'] as String),
      endTime: DateTime.parse(json['endTime'] as String),
      duration: Duration(seconds: json['duration'] as int),
      dances: (json['dances'] as List<dynamic>).cast<String>(),
      info: (json['info'] as List<dynamic>)
          .map((i) => EventInfo.fromJson(i as Map<String, dynamic>))
          .toList(),
      parts: (json['parts'] as List<dynamic>)
          .map((p) => EventPart.fromJson(p as Map<String, dynamic>))
          .toList(),
      isFavorite: json['isFavorite'] as bool? ?? false,
      isPast: json['isPast'] as bool? ?? false,
      badge: json['badge'] as String?,
    );
  }

  @override
  List<Object?> get props => [
        id,
        title,
        description,
        organizer,
        venue,
        startTime,
        endTime,
        duration,
        dances,
        info,
        parts,
        isFavorite,
        isPast,
        badge,
      ];
}

File: backend/dancee_events/internal/models/event.go
package models

// Address represents a physical address
type Address struct {
	Street     string `json:"street" firestore:"street"`
	City       string `json:"city" firestore:"city"`
	PostalCode string `json:"postalCode" firestore:"postalCode"`
	Country    string `json:"country" firestore:"country"`
}

// Venue represents an event venue
type Venue struct {
	Name        string   `json:"name" firestore:"name"`
	Address     Address  `json:"address" firestore:"address"`
	Description string  `json:"description" firestore:"description"`
	Latitude    float64 `json:"latitude" firestore:"latitude"`
	Longitude   float64 `json:"longitude" firestore:"longitude"`
}

// EventInfo represents additional event information
type EventInfo struct {
	Type  string `json:"type" firestore:"type"`
	Key   string `json:"key" firestore:"key"`
	Value string `json:"value" firestore:"value"`
}

// EventPart represents a part/segment of an event
type EventPart struct {
	Name        string   `json:"name" firestore:"name"`
	Description *string  `json:"description,omitempty" firestore:"description,omitempty"`
	Type        string   `json:"type" firestore:"type"`
	StartTime   string   `json:"startTime" firestore:"startTime"`
	EndTime     *string  `json:"endTime,omitempty" firestore:"endTime,omitempty"`
	DJs         []string `json:"djs,omitempty" firestore:"djs,omitempty"`
	Lectors     []string `json:"lectors,omitempty" firestore:"lectors,omitempty"`
}

// Event represents a dance event
type Event struct {
	ID          string       `json:"id" firestore:"-"`
	Title       string       `json:"title" firestore:"title"`
	Description string       `json:"description" firestore:"description"`
	Organizer   string       `json:"organizer" firestore:"organizer"`
	Venue       Venue        `json:"venue" firestore:"venue"`
	StartTime   string       `json:"startTime" firestore:"startTime"`
	EndTime     *string      `json:"endTime,omitempty" firestore:"endTime,omitempty"`
	Duration    *int64       `json:"duration,omitempty" firestore:"duration,omitempty"`
	Dances      []string     `json:"dances" firestore:"dances"`
	Info        []EventInfo  `json:"info,omitempty" firestore:"info,omitempty"`
	Parts       []EventPart  `json:"parts,omitempty" firestore:"parts,omitempty"`
	IsFavorite  *bool        `json:"isFavorite,omitempty" firestore:"isFavorite,omitempty"`
	IsPast      *bool        `json:"isPast,omitempty" firestore:"-"`
}

// AddFavoriteRequest represents the request to add a favorite
type AddFavoriteRequest struct {
	UserID  string `json:"userId" binding:"required"`
	EventID string `json:"eventId" binding:"required"`
}

// CreateEventRequest represents the request body for creating/updating an event.
// Uses Gin's binding tags for required field validation.
type CreateEventRequest struct {
	Title       string      `json:"title" binding:"required"`
	Description string      `json:"description" binding:"required"`
	Organizer   string      `json:"organizer" binding:"required"`
	Venue       Venue       `json:"venue" binding:"required"`
	StartTime   string      `json:"startTime" binding:"required"`
	EndTime     *string     `json:"endTime,omitempty"`
	Duration    *int64      `json:"duration,omitempty"`
	Dances      []string    `json:"dances" binding:"required,min=1"`
	Info        []EventInfo `json:"info,omitempty"`
	Parts       []EventPart `json:"parts,omitempty"`
}

// ToEvent converts a CreateEventRequest to an Event model.
func (r *CreateEventRequest) ToEvent() *Event {
	return &Event{
		Title:       r.Title,
		Description: r.Description,
		Organizer:   r.Organizer,
		Venue:       r.Venue,
		StartTime:   r.StartTime,
		EndTime:     r.EndTime,
		Duration:    r.Duration,
		Dances:      r.Dances,
		Info:        r.Info,
		Parts:       r.Parts,
	}
}